### PR TITLE
1190 android loader and splash

### LINKED ISSUE #1190: Android fails to load or display loader
## What went wrong? 😲
 
I'm wondering if we can prevent this from showing up as much as it does?
This is more of a discussion point / food for thought. Any ideas appreciated

Screenshot 2023-02-02 at 9 21 42 AM

This shows fairly often for me
1. briefly on start - when it would be good to have a loader instead
2. sometimes though, it remains. even when the server is available. restarting the app sometimes works fine 🤷 
3. when there's an error, as in the example above

## Expected behaviour 🤔
- A friendly error screen, if there's an error. 
- A loading screen when the app is loading content or waiting for the server
- A retry perhaps.. if the server isn't currently available, pause and try again, in order to prevent scenario three above

## How to Reproduce 🔨

Try the android app for a while and see if this is a valid issue, or just me..

## Your environment 🌱
 - Version 1.1.3
 - Platform android

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1190 

# 👩🏻‍💻 What does this PR do? 
 
This PR does two main things: wait for the local server to be available before loading a page from it, and retry until ready. This is particularly a problem after clearing the app data, when the server takes just slightly longer to start and shows an error. you need to close / refresh the app to get the oms app to show; and secondly hides the jumping around between screens, until the app is ready for you. The changes are:

- Adds a new splash screen, which is then hidden by the `useNativeClient` hook when the hook is loaded
- Retry the server url loading 5 times, with a 1 second delay after each
- Adds an error page, shown when the server url cannot be loaded, there's an error or the server isn't responding

I started with the `@capacitor\splash-screen` plugin but had much difficulty getting that to work. It simply would not work for me from the client - I could add a custom splash though I had to upgrade to a min SDK version of 31 for that to work. Eventually, I just implemented the splash manually. 

Am unable to completely remove the initial splash screen, so there's a double splash - looks better if we upgrade to a minSDK of 32 but I like that we support earlier android levels, so went back to a clumsiness.

This is how the startup looks now:

https://user-images.githubusercontent.com/9192912/220519892-5d7a01a6-a338-4781-b05e-e36737b7a10d.mp4

The error screen looks like this:

Screenshot 2023-02-23 at 9 00 24 AM

# 🧪 How has/should this change been tested? 

Manually testing. I have a tablet if you want to use it. Have only tested on the samsung and lenovo tablets.

## 📃 Documentation

- [ ] Need to document all of the android startup bits

### CODE REVIEW

File: client/packages/android/app/src/main/java/org/openmsupply/client/ErrorPage.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.openmsupply.client;
+
+public class ErrorPage {
+    // the unhappy man svg and the text "Oops! Something's gone wrong"
```
Comment (andreievg) [inline client/packages/android/app/src/main/java/org/openmsupply/client/ErrorPage.java:4]: Would be great to have some instructions of how this was made, and what about splash screen can splash screen be encoded this way ?

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 2]
```diff
@@ -64,12 +78,120 @@ public String getLocalUrl() {
...
+                    Drawable splash = getSplashDrawable(context);
+                    LinearLayout parent = new LinearLayout(context);
+                    parent.setLayoutParams(
+                            new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT,
+                                    LinearLayout.LayoutParams.WRAP_CONTENT));
+                    parent.setOrientation(LinearLayout.VERTICAL);
+                    if (splash != null) {
+                        parent.setBackground(splash);
+                    }
+                    this.splashScreen.setContentView(parent);
```
Comment (andreievg) [inline client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:168]: can simplify this slightly with:

https://github.com/openmsupply/open-msupply/compare/1190-example1...1190-example2

Although I kind of like the encoding better, and it also solves the problem with flicker i described (I think)

**Review by andreievg — CHANGES_REQUESTED** [review]
I was trying to investigate the reason why `initialisation` screen pops up first, so did a little bit of debuggin:

```diff
+const Check: React.FC<PropsWithChildrenOnly> = ({ children }) => {
+  const location = useLocation();
+  console.log('here we are', JSON.stringify(location, null, ' '));
+  return <> {children} </>;
+};
+
 const Host = () => (
   <React.Suspense fallback={<div />}>
     <IntlProvider>
@@ -76,6 +84,7 @@ const Host = () => (
                         <AuthenticationAlert />
                         <Viewport>
                           <Box display="flex" style={{ minHeight: '100%' }}>
+                            <Check>
                               <Routes>
                                 <Route
                                   path={RouteBuilder.create(
@@ -103,6 +112,7 @@ const Host = () => (
                                 />
                                 <Route path="*" element={<Site />} />
                               </Routes>
+                            </Check>
                           </Box>
                         </Viewport>
                       </BrowserRouter>
                  
```

Which resulted in:

```
2023-03-01 11:59:31.541 11934-11934 Capacitor/Console       org.openmsupply.client               I  File: https://localhost:8000/455a51d959a646054dcf.js - Line 250 - Msg: here we are {
                                                                                                     "pathname": "/",
                                                                                                     "search": "",
                                                                                                     "hash": "",
                                                                                                     "state": null,
                                                                                                     "key": "default"
                                                                                                    }
2023-03-01 11:59:31.616 11934-11934 Capacitor/Console       org.openmsupply.client               I  File: https://localhost:8000/455a51d959a646054dcf.js - Line 250 - Msg: here we are {
                                                                                                     "pathname": "/login",
                                                                                                     "search": "",
                                                                                                     "hash": "",
                                                                                                     "state": {
                                                                                                      "from": {
                                                                                                       "pathname": "/",
                                                                                                       "search": "",
                                                                                                       "hash": "",
                                                                                                       "state": null,
                                                                                                       "key": "default"
                                                                                                      }
                                                                                                     },
                                                                                                     "key": "1spamblj"
                                                                                                    }
2023-03-01 11:59:31.788 11934-11934 Capacitor/Console       org.openmsupply.client               I  File: https://localhost:8000/455a51d959a646054dcf.js - Line 250 - Msg: here we are {
                                                                                                     "pathname": "/initialise",
                                                                                                     "search": "",
                                                                                                     "hash": "",
                                                                                                     "state": null,
                                                                                                     "key": "v6ab7cr4"
                                                                                                    }
2023-03-01 11:59:32.196 11934-11934 Capacitor/Console       org.openmsupply.client               I  File: https://localhost:8000/455a51d959a646054dcf.js - Line 250 - Msg: here we are {
                                                                                                     "pathname": "/android",
                                                                                                     "search": "",
                                                                                                     "hash": "",
                                                                                                     "state": null,
                                                                                                     "key": "default"
                                                                                                    }
```

This is because url in `capacitor.config.ts` is set to `https://localhost:8000`, and it load in the background and clears splashscreen before below is run: https://github.com/openmsupply/open-msupply/blob/96a78d248fc96bfc5ed04adf6a2d9651c8bfafa7/client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java#L125-L126

After changing `https://localhost:8000` to `https://localhost:8000/android` and making sure `splashscreen` is only removed when we navigated to our final destination, it seems to fix the issue:

https://github.com/openmsupply/open-msupply/compare/1190-android-loader-and-splash...1190-example1

There was still a little flicker remaining, since behind the sceen there was still a call to `https://localhost:8000/android` at the start (through the url preference), and the we do navigation there manually: https://github.com/openmsupply/open-msupply/blob/96a78d248fc96bfc5ed04adf6a2d9651c8bfafa7/client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java#L125-L126

We can improve this by only navigating there if splash screen is still present. 

Main change request for this PR is the reduction of flickering as per above, otherwise happy to approve, other improvements are :nit.

Just before I finished the review I saw the encodingHTML done for `oops`, can this be done for splash screen too ? And then just simply on start navigate to that splash screen, if this can be done we dont' need any of the show/hide splash screen logic in android or front end code etc.. and I think it would make it much simpler ?

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 3]
```diff
@@ -64,12 +78,120 @@ public String getLocalUrl() {
...
+                }
+            }
+        });
+        thread.start();
+    }
+
+    private Drawable getSplashDrawable(Context context) {
+        int splashId = context.getResources().getIdentifier("splash", "drawable", context.getPackageName());
+        try {
+            Drawable drawable = context.getResources().getDrawable(splashId, context.getTheme());
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:138]: perfect! thanks

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 4]
```diff
@@ -64,12 +78,120 @@ public String getLocalUrl() {
...
+                        urlc.connect();
+                        if (urlc.getResponseCode() == 200) {
+                            isServerRunning = true;
+                        }
+                    } catch (SSLHandshakeException e) {
+                        // server is running and responding with an SSL error
+                        // which we will ignore, so ok to proceed
+                        isServerRunning = true;
+                    } catch (Exception e) {
+                        Log.e("omSupply", e.getMessage());
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:116]: didn't see anything - figured it's a low priority so have put in a CONST. I'm not expecting these logs to be used / useful

File: client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java [hunk 5]
```diff
@@ -64,12 +78,120 @@ public String getLocalUrl() {
...
+        });
+        thread.start();
+    }
+
+    private Drawable getSplashDrawable(Context context) {
+        int splashId = context.getResources().getIdentifier("splash", "drawable", context.getPackageName());
+        try {
+            Drawable drawable = context.getResources().getDrawable(splashId, context.getTheme());
+            return drawable;
+        } catch (Resources.NotFoundException ex) {
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/android/app/src/main/java/org/openmsupply/client/NativeApi.java:140]: couldn't find anything 🤷

**Comment by mark-prins** [comment]
🙏  thanks @andreievg - see how the encoded version of the splash looks for you (96a78d248)
Working nicely for me. Have also changed the encoded HTML to actual HTML to make the code clearer and easier to change.

File: client/packages/android/README.MD [hunk 6]
```diff
@@ -88,18 +88,41 @@ For this reason `ExtendedWebViewClient` was created to manually inject the requi
...
 
-`app/src/main/assets/public` directory is typically not commited, but since it's only going to have cordova artifacts, it is in our case (to reduce setup). 
+`app/src/main/assets/public` directory is typically not committed, but since it's only going to have cordova artifacts, it is in our case (to reduce setup). 
 
 `npx cap copy` or `yarn apply-config` should only be run when cordova plugins are added or updated or when updating capacitor.config.ts
 
+### Plugins
+
+When adding plugins, the standard process (`npx cap sync` or `yarn apply-config`) is not working and I've had to manually add them:
```
Comment (andreievg) [inline client/packages/android/README.MD:105]: Thank you Mark, i had problems with `npx cap sync` also, with some dependencies being added (this is even before the refactor to discovery), I thought I fixed it in the changes but nope, adding plugins was still not working (found out not long ago but never fixed). Your instructions look very good

**Review by andreievg — APPROVED** [review]
Thanks Mark, works very nice !

Is that `plugin->SplashScreen` in capacitor.config.ts still required ?

File: client/packages/android/app/src/main/java/org/openmsupply/client/SplashPage.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package org.openmsupply.client;
+
+import android.util.Base64;
+
+public class SplashPage {
+    // Simple HTML to show the splash image centred on the page
+    static String html = "<html>\n" +
```
Comment (andreievg) [inline client/packages/android/app/src/main/java/org/openmsupply/client/SplashPage.java:7]: Thanks, very clear now

**Comment by mark-prins** [comment]
> Thanks Mark, works very nice !
> 
> Is that `plugin->SplashScreen` in capacitor.config.ts still required ?

ooh. nope.

**Pull request merged**